lib/uksgx/include/uk/sgx_cpu.h

```diff
@@ -0,0 +1,283 @@
+	" .popsection\n"
+
+#define _ASM_EXTABLE(from, to)					\
+	_ASM_EXTABLE_TYPE(from, to, EX_TYPE_DEFAULT)
+
+/* Referred to as IA32_FEATURE_CONTROL in Intel's SDM. */
+#define X86_MSR_IA32_FEAT_CTL		            0x0000003a
+#define X86_FEAT_CTL_LOCKED				        BIT(0)
+#define X86_FEAT_CTL_VMX_ENABLED_INSIDE_SMX		BIT(1)
+#define X86_FEAT_CTL_VMX_ENABLED_OUTSIDE_SMX	BIT(2)
+#define X86_FEAT_CTL_SGX_LC_ENABLED			    BIT(17)
+#define X86_FEAT_CTL_SGX_ENABLED			    BIT(18)
+#define X86_FEAT_CTL_LMCE_ENABLED			    BIT(20)
+
+#define X86_MSR_IA32_EFER			            0xC0000080 /* extended feature enables */
+#define X86_EFER_SYSCALL			    		0x1UL
+#define X86_MSR_IA32_STAR			            0xC0000081 /* legacy mode SYSCALL target */
+#define X86_MSR_IA32_LSTAR			            0xC0000082 /* long mode SYSCALL target */
+#define X86_MSR_IA32_FMASK			            0xC0000084 /* legacy mode SYSCALL mask */
+
+/* Intel SGX Launch Enclave Public Key Hash MSRs */
+#define MSR_IA32_SGXLEPUBKEYHASH0	0x0000008C
+#define MSR_IA32_SGXLEPUBKEYHASH1	0x0000008D
+#define MSR_IA32_SGXLEPUBKEYHASH2	0x0000008E
+#define MSR_IA32_SGXLEPUBKEYHASH3	0x0000008F
+
+/* __cpuid_count(unsinged int info[4], unsigned int leaf, unsigned int subleaf); */
+#define __cpuid_count(x, y, z)                                                       \
+	asm volatile("cpuid"                                                   \
+		     : "=a"(x[0]), "=b"(x[1]), "=c"(x[2]), "=d"(x[3])          \
+		     : "a"(y), "c"(z))
+
+/* __cpuid(unsinged int info[4], unsigned int leaf); */
+#define __cpuid(x, y)                                                    \
+	asm volatile("cpuid"                                                   \
+		     : "=a"(x[0]), "=b"(x[1]), "=c"(x[2]), "=d"(x[3])          \
+		     : "a"(y))
+
+#define Genu 0x756e6547
+#define ineI 0x49656e69
+#define ntel 0x6c65746e
+
+static inline void rdmsr(unsigned int msr, __u32 *lo, __u32 *hi)
+{
+	asm volatile("rdmsr" : "=a"(*lo), "=d"(*hi)
+			     : "c"(msr));
+}
+
+static inline __u64 rdmsrl(unsigned int msr)
+{
+	__u32 lo, hi;
+
+	rdmsr(msr, &lo, &hi);
+	return ((__u64) lo | (__u64) hi << 32);
+}
+
+/* wrmsr with exception handling */
+static inline void wrmsr(unsigned int msr, __u32 low, __u32 high)
+{
+	asm volatile("wrmsr"
+			     : /* no outputs */
+			     : "c"(msr), "a"(low), "d"(high));
+}
+
+/*
+ * 64-bit version of wrmsr():
+ */
+static inline void wrmsrl(__u32 msr, __u64 val)
+{
+	wrmsr(msr, (__u32)val,  (__u32)(val >> 32));
+}
+
+
+/*
+ * Volatile isn't enough to prevent the compiler from reordering the
+ * read/write functions for the control registers and messing everything up.
+ * A memory clobber would solve the problem, but would prevent reordering of
+ * all loads stores around it, which can hurt performance. Solution is to
+ * use a variable and mimic reads and writes to it to enforce serialization
+ */
+extern unsigned long __force_order;
+
+static inline unsigned long native_read_cr0(void)
```

Comment:
Maybe it would be a good idea to define the functions related to the x86 architecture in `plat/common/include/x86/cpu.h`, so we don't have duplicates

Reply:
Yep, I defined them in my own lib because at first I am not familiar with other part of Unikraft hence do not want to modify it. I will perform a simplify and clarify once we reach a great milestone of this project, say make a "hello world" work from SGX.